### ISSUE TITLE
Computed Properties aren't bound correctly during Object/Class evaluation

### ISSUE THREAD

**Issue by joeywatts** [microsoft/TypeScript#27864]
**TypeScript Version:**  3.2.0-dev.20181011

**Search Terms:** computed property expression

**Code**

```ts
const classes = [];
for (let i = 0; i <= 10; ++i) {
  classes.push(
    class A {
      [i] = "my property";
    }
  );
}
for (const clazz of classes) {
  console.log(Object.getOwnPropertyNames(new clazz()));
}
```

**Expected behavior:** The log statements indicate that each class in `classes` has a different property name (`i` should be evaluated at the time of the class evaluation and all instances of that class should have a property name corresponding to that evaluation of `i`).

**Actual behavior:** Compiled code logs:

> [ '10' ]
> [ '10' ]
> [ '10' ]
> [ '10' ]
> [ '10' ]
> [ '10' ]
> [ '10' ]
> [ '10' ]
> [ '10' ]
> [ '10' ]
> [ '10' ]

[**Playground Link**](https://www.typescriptlang.org/play/index.html#src=const%20classes%20%3D%20%5B%5D%3B%0D%0Afor%20(let%20i%20%3D%200%3B%20i%20%3C%3D%2010%3B%20%2B%2Bi)%20%7B%0D%0A%20%20classes.push(%0D%0A%20%20%20%20class%20A%20%7B%0D%0A%20%20%20%20%20%20%5Bi%5D%20%3D%20%22my%20property%22%3B%0D%0A%20%20%20%20%7D%0D%0A%20%20)%3B%0D%0A%7D%0D%0Afor%20(const%20clazz%20of%20classes)%20%7B%0D%0A%20%20console.log(Object.getOwnPropertyNames(new%20clazz()))%3B%0D%0A%7D)

**Comment by mheiber** [comment]
There's a type error on the line with the computed property:

> A computed property name in a class property declaration must refer to an expression whose type is a literal type or a 'unique symbol' type.

However
- It's still possible to produce correct emit in this case, even though there's a type error (if that's desirable)
- The [class fields proposal](https://github.com/tc39/proposal-class-fields/issues) allows code like in Joey's example, and  Chrome Canary run's Joey's example with the correct output.

**Comment by DanielRosenwasser** [comment]
Taking PRs and marking it as moderate (assuming you are familiar with spec-reading and the transform pipeline).

**Comment by mheiber** [comment]
What are the implications of getting rid of the type error? Will allowing arbitrary string and symbol values in computed field declarations affect inference?

**Comment by mheiber** [comment]
Maybe we can lean on the iife stuff that's already done to transpile `let` and `const`.

The currently generated code for Joey's example when targeting ES2015 is:

```js
var _a, _b;
"use strict";
const classes = [];
for (let i = 0; i <= 10; ++i) {
    classes.push((_b = class A {
            constructor() {
                this[_a] = "my property";
            }
        },
        _a = i,
        _b));
}
for (const clazz of classes) {
    console.log(Object.getOwnPropertyNames(new clazz()));
}
```

If we move `_a` and `_b` into the scope of the `for` loop and use `let` then the output is correct:

```js
var _a, _b;
"use strict";
const classes = [];
for (let i = 0; i <= 10; ++i) {
    classes.push((_b = class A {
            constructor() {
                this[_a] = "my property";
            }
        },
        _a = i,
        _b));
}
for (const clazz of classes) {
    console.log(Object.getOwnPropertyNames(new clazz()));
}
```

**Comment by Kingwl** [comment]
> If we move _a and _b into the containing scope of the class and use let then the output is correct:  

but `ForStatement` does not has a lexical scope

**Comment by Kingwl** [comment]
maybe inject template variable into IIFE is better? I'll take a try on it

**Comment by Kingwl** [comment]
the special temporary variable and hoist is happened in ts transformer before es2015
perhaps we need a block level hoist?

**Comment by mheiber** [comment]
@Kingwl agreed. we're trying that approach in https://github.com/joeywatts/TypeScript/pull/15

**Comment by joeywatts** [comment]
I noticed that there's also an issue with the binding of property declaration initializers in class expressions which are within an iteration statement. For example:

```ts
let arr = [];
for (let i = 0; i < 5; ++i) {
    class C {
        test = i
    }
    arr.push(new C());
}
arr.forEach(obj => console.log(obj.test));
```

Since the `i` is block scoped within the loop, each class should capture its own value of `i` (causing it to print 0, 1, 2, 3, 4). The TypeScript compiler currently does not preserve this binding when targeting ES5 or lower. This currently produces no compile error and leads to different run time behavior when targeting ES2015 vs ES5.

I've investigated the changes required to the checker in order to pick up this binding, and it appears to me that a simple change to the `checkNestedScopeBinding` function to account for usage sites within instance property initializers (in addition to functions). Should I open a separate issue for this?

**Comment by xialvjun** [comment]
I think code like this should be supported:
```ts
class A<P extends PropertyKey> {
    [p in P]: number;
}
```

But now it only support:
```ts
const m = Symbol('m');
const n = 'n' as 'n';

class A {
  [m]: number;
  [n]: string;
}
```

**Comment by RyanCavanaugh** [comment]
Output is correct now
